jsonrpc/websocket.go

```diff
@@ -54,16 +54,21 @@ func (ws *Websocket) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// TODO include connection information, such as the remote address, in the logs.
 
-	wsc := newWebsocketConn(r.Context(), conn, ws.connParams, ws.listener)
+	wsc := newWebsocketConn(r.Context(), conn, ws.connParams)
 
 	ctx := context.WithValue(wsc.ctx, ConnKey{}, wsc)
 	for {
+		var msg []byte
+		_, msg, err = wsc.conn.Read(ctx)
```

Comment:
We will allocate a buffer here for each request, on top of the `io.Reader` and `bufio.Reader` allocations that server will do.

I am not sure if this is the most efficient approach.

Reply:
I dont think so. 

These approaches also break the bidirectional communication that we introduced in the last PR, no? But I think bidirectional communication between handler and client was a mistake.  Handler should only be able to write to the connection, it shouldnt be able to read from it. Reading should be only available to `jsonrpc.Server`.

Something like would work better IMO.

``` golang
bufferedWSConn := bufio.Reader(wsc.conn)
writerCtx := context.WithValue(key, wsc.conn)
for {
   resp, err := ws.rpc.HandleReader(ctx, reader) 
   if err != nil { break }
   .
   .
   .
}
```

Reply:
With the `bufio` approach, `wsc.conn` must be an `io.Reader`. What do you think the `wsc.conn.Read` method should look like?

> But I think bidirectional communication between handler and client was a mistake.

I agree. I changed the return type of `jsonrpc.ConnFromContext` from `io.ReadWriter` to `io.Writer` before merging (just realized I forgot to update the commit message, unfortunately).

Reply:
> . I changed the return type of jsonrpc.ConnFromContext from io.ReadWriter to io.Writer before merging
Nice, I didnt realize that.

> With the bufio approach, wsc.conn must be an io.Reader
Looks like we can just use this? https://pkg.go.dev/nhooyr.io/websocket#NetConn

Reply:
> Looks like we can just use this? https://pkg.go.dev/nhooyr.io/websocket#NetConn

Oh cool, I forgot about that. There is a bug in that function, so we can use Tailscale's workaround I think: https://github.com/tailscale/tailscale/pull/5983